# Exploring eBay Car Sales Data

In this project, we will clean the data and analyze the used car listings in dataset, [auto.csv](), extracted from [here](https://www.kaggle.com/orgesleka/used-cars-database/data). The dataset has 50,000 data points including 20 columns and below is the data dictionary/fields of the `autos.csv` dataset:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

In [1]:
# importing libraries

import pandas as pd
import numpy as np

In [2]:
# reading file into pandas
# default encoder `UTF-8` caused encoding error therefore used
# `Latin-1` encoding

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
# extracting first and last few rows of the dataset

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
# displaying info for the dataset

autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

The `autos` dataframe contains 20 columns, most of which have dtypes of *object* (strings) with the exceptions of five columns; `yearOfRegistration`, `powerPS`, `monthOfRegistration`, `nrOfPictures` and `postalCode` that have dtypes of *int64*. Some columns have *null values* and the column names use *camelcase* instead of python's preferred *snakecase*, which mean we can't just replace spaces with underscores. Also three columns; `dateCrawled`, `dateCreated` and `lastSeen` contain date & time but have dtypes of *object*.

In [5]:
# printing an array of the dataset's column names

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# copying an array to edit the column names
old_new_labels = {'yearOfRegistration': 'registration_year',
                 'monthOfRegistration': 'registration_month', 
                 'notRepairedDamage': 'unrepaired_damage', 
                 'dateCreated': 'ad_created', 
                 'dateCrawled': 'date_crawled', 
                 'name': 'car_name', 
                 'seller': 'private_seller_dealer', 
                 'offerType': 'listing_type', 
                 'price': 'sale_price_usd', 
                 'abtest': 'ab_test', 
                 'vehicleType': 'vehicle_type', 
                 'gearbox': 'gear_box', 
                 'powerPS': 'power_ps', 
                 'model': 'car_model', 
                 'odometer': 'odometer_km', 
                 'fuelType': 'fuel_type', 
                 'brand': 'car_brand', 
                 'nrOfPictures': 'no_of_pictures', 
                 'postalCode': 'postal_code',
                 'lastSeen': 'last_seen'}

autos = autos.rename(columns=old_new_labels)

autos.head()                     

date_crawled                                           car_name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

  private_seller_dealer listing_type sale_price_usd  ab_test vehicle_type  \
0                privat      Angebot         $5,000  control          bus   
1                privat      Angebot         $8,500  control    limousine   
2                privat      Angebot         $8,990     test    limousine   
3                privat      Angebot         $4,350  control   kleinwagen   
4                privat      Angebot         $1,350     test        kombi   

   registration_year   gear_box  power_ps car_model odometer_km  \
0               2004    manuell       158    andere   150,000km   
1               1997  automatik       286       7er   150,000km   
2               2009    manuell       102      golf    70,000km   
3               2007  automatik        71    fortwo    70,000km   
4               2003    manuell         0     focus   150,000km   

   registration_month fuel_type   car_brand unrepaired_damage  \
0                   3       lpg     peugeot              nein   
1                   6    benzin         bmw              nein   
2                   7    benzin  volkswagen              nein   
3                   6    benzin       smart              nein   
4                   7    benzin        ford              nein   

            ad_created  no_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00               0        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00               0        35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00               0        33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00               0        39218  2016-04-01 14:38:50

I have changed the columns labels from *camelcase* to *snakecase* and reworded few columns names to make it more readable.

In [7]:
autos.describe(include='all')

date_crawled     car_name private_seller_dealer listing_type  \
count                 50000        50000                 50000        50000   
unique                48213        38754                     2            2   
top     2016-03-27 22:55:05  Ford_Fiesta                privat      Angebot   
freq                      3           78                 49999        49999   
mean                    NaN          NaN                   NaN          NaN   
std                     NaN          NaN                   NaN          NaN   
min                     NaN          NaN                   NaN          NaN   
25%                     NaN          NaN                   NaN          NaN   
50%                     NaN          NaN                   NaN          NaN   
75%                     NaN          NaN                   NaN          NaN   
max                     NaN          NaN                   NaN          NaN   

       sale_price_usd ab_test vehicle_type  registration_year gear_box  \
count           50000   50000        44905       50000.000000    47320   
unique           2357       2            8                NaN        2   
top                $0    test    limousine                NaN  manuell   
freq             1421   25756        12859                NaN    36993   
mean              NaN     NaN          NaN        2005.073280      NaN   
std               NaN     NaN          NaN         105.712813      NaN   
min               NaN     NaN          NaN        1000.000000      NaN   
25%               NaN     NaN          NaN        1999.000000      NaN   
50%               NaN     NaN          NaN        2003.000000      NaN   
75%               NaN     NaN          NaN        2008.000000      NaN   
max               NaN     NaN          NaN        9999.000000      NaN   

            power_ps car_model odometer_km  registration_month fuel_type  \
count   50000.000000     47242       50000        50000.000000     45518   
unique           NaN       245          13                 NaN         7   
top              NaN      golf   150,000km                 NaN    benzin   
freq             NaN      4024       32424                 NaN     30107   
mean      116.355920       NaN         NaN            5.723360       NaN   
std       209.216627       NaN         NaN            3.711984       NaN   
min         0.000000       NaN         NaN            0.000000       NaN   
25%        70.000000       NaN         NaN            3.000000       NaN   
50%       105.000000       NaN         NaN            6.000000       NaN   
75%       150.000000       NaN         NaN            9.000000       NaN   
max     17700.000000       NaN         NaN           12.000000       NaN   

         car_brand unrepaired_damage           ad_created  no_of_pictures  \
count        50000             40171                50000         50000.0   
unique          40                 2                   76             NaN   
top     volkswagen              nein  2016-04-03 00:00:00             NaN   
freq         10687             35232                 1946             NaN   
mean           NaN               NaN                  NaN             0.0   
std            NaN               NaN                  NaN             0.0   
min            NaN               NaN                  NaN             0.0   
25%            NaN               NaN                  NaN             0.0   
50%            NaN               NaN                  NaN             0.0   
75%            NaN               NaN                  NaN             0.0   
max            NaN               NaN                  NaN             0.0   

         postal_code            last_seen  
count   50000.000000                50000  
unique           NaN                39481  
top              NaN  2016-04-07 06:17:27  
freq             NaN                    8  
mean    50813.627300                  NaN  
std     25779.747957                  NaN  
min      1067.000000           

After looking at the above results of descriptive statistics, `no_of_pictures` column has null values so they need to be dropped. Morever, we need to look further into `registration_year` column as it shows maximum value as 9999 which is inaccurate. Other columns `sale_price_USD` and `odometer_km` are numeric data storedas text so they need to be cleaned and converted to numeric dtype (int or float). There are a number of text columns where nearly all of the values are the same; `privat_seller_dealer` and `listing_type`. However, `no_of_pictures` column also seems odd, we'll examine it further.

In [8]:
autos['no_of_pictures'].value_counts()

0    50000
Name: no_of_pictures, dtype: int64

In [9]:
# Above output confirms that `no_of_pictures' column has 0 for every column.
# We'll drop this column, plus the other two, `private_seller_dealer` and
# `listing_type`

autos = autos.drop(['private_seller_dealer', 
                    'listing_type', 
                    'no_of_pictures'],
                   axis=1)

In [10]:
# examining non-numeric values of `sales_price` and `odometer_km`columns

autos[['sale_price_usd', 'odometer_km']].head()

sale_price_usd odometer_km
0         $5,000   150,000km
1         $8,500   150,000km
2         $8,990    70,000km
3         $4,350    70,000km
4         $1,350   150,000km

In [11]:
# converting `sale_price_usd` column from text dtype to numeric dtype (int)

autos['sale_price_usd'] = (autos['sale_price_usd']
                       .str.replace(',', '')
                       .str.replace('$', '')
                       .astype(int)
                      )

In [12]:
# examining first five elements and dtype of series

autos.sale_price_usd.head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: sale_price_usd, dtype: int32

In [13]:
# converting `odometer_km` column from text dtype to numeric dtype (int)

autos.odometer_km = (autos.odometer_km
                     .str.replace('km', '')
                     .str.replace(',', '')
                     .astype(int)
                     )

In [14]:
# examining first five elements and dtype of series

autos.odometer_km.head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32

In [15]:
# analyzing `sale_price_USD` and `odometer_km` to determine outliers

# shape of series with unique non-null values
print('Shape of sale_price_usd:', autos.sale_price_usd.unique().shape)
print('Shape of odometer_km:', autos.odometer_km.unique().shape)

Shape of sale_price_usd: (2357,)
Shape of odometer_km: (13,)


In [16]:
# descriptive stats summary for both series
print(autos['sale_price_usd'].describe())
print('\n')
print(autos['odometer_km'].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: sale_price_usd, dtype: float64


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [17]:
autos['sale_price_usd'].value_counts().head(20)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: sale_price_usd, dtype: int64

In [18]:
autos[autos['sale_price_usd'] == 0]

date_crawled                                           car_name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146    2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
226    2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
248    2016-03-08 22:57:53                             VW_Passat_zu_verkaufen   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
301    2016-03-08 20:37:59                         Kaufe_alle_Autos_bietet_an   
323    2016-03-15 16:54:59                                W220_Unfallfahrzeug   
327    2016-04-03 13:49:21              BMW_Cabrio_318i_/_BMW_318_tds_Touring   
347    2016-04-02 11:48:02                                  Volvo_TP21__Sugga   
366    2016-03-21 20:36:25  Ford_Escort_Express_75__LKW_Zulussung___Diesel...   
389    2016-03-08 11:38:52                                    Vw_polo_classic   
418    2016-03-29 14:43:24         Fiat_SCUDO_8_Sitzer_Bus__Diesel_JTD__80_KW   
427    2016-03-12 11:44:47                 TOP_ZUSTAND_GrunePlakette_Euro4!!!   
430    2016-03-18 23:52:40                                  Winterraeder_FORD   
449    2016-04-03 20:59:59                           Tausche_gegen_ein_smart.   
487    2016-03-14 21:41:36  MAG_MICH_KEINER_HABEN_Opel_Omega_B_Caravan_2.0...   
502    2016-03-25 23:55:24                               C32_AMG_C55_C43_w203   
519    2016-03-28 08:53:58  Verkaufe/Tausche_Seat_Inca__VW_Caddy__aus_2002...   
565    2016-03-20 20:49:36     Tausche_Audi_S3_8p_gegen_Golf_VI_Gti_mit_xenon   
593    2016-04-01 21:44:15                              CLK_200__BRABUS_Optik   
660    2016-03-29 21:44:40                                       Opel_Corsa_B   
661    2016-03-18 22:49:48               Renault_Megane_Scenic_1.6_Tuev_12/17   
...                    ...                                                ...   
48882  2016-03-14 23:58:47                                         SUCHE_AUTO   
48889  2016-04-01 10:50:52                           Audi_A4_____TueV_09.2016   
48891  2016-03-20 22:50:14        MB_E_220_CDI__TAUSCHEN_nur_SUV/GELÄNDEWAGEN   
48982  2016-03-08 12:48:33                           Bmw_525D_Sport_Pro..Navi   
48985  2016-03-27 08:55:25                             Stockcar_Autos_Peugeot   
49043  2016-03-07 11:48:55                              Opel_Vectra_B_Caravan   
49082  2016-03-13 17:06:58         Passat_35i_b4_in_candy_weiss_original_lack   
49254  2016-03-21 10:49:57                           Corrado_G60_mit_2e_Motor   
49301  2016-03-20 20:52:40                           Golf_3_Variant_Kombi_1.8   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49350  2016-03-19 23:57:15                                     Golf_II_Diesel   
49404  2016-04-04 03:03:16                                Golf_2_BBS_LEDER_KW   
49408  2016-03-30 20:57:06                                         OPEL_ASTRA   
49479  2016-04-01 23:55:01                            Tausch_mit_frischen_TÜV   
49496  2016-03-26 13:55:28                                        Bmw_e39_520   
49504  2016-03-13 19:42:57                                 Meriva_Lpg_1_6_16v   
49507  2016-03-16 17:38:32                 BMW_528i_E39_Schaltgetriebe_Unfall   
49525  2016-03-09 17:59:02              Bmw_e

There are 1,421 cars listed with $0 sales_price, given that this is only 2% of the of the cars, we may consider removing these rows. The maximum price is one hundred million dollars, which is quite abnormal and needs to be looked into further.

In [19]:
autos['sale_price_usd'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: sale_price_usd, dtype: int64

In [20]:
autos['sale_price_usd'].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: sale_price_usd, dtype: int64

By looking at above results, we may say that over \$1 million sale_price of cars is unrealistic and therefore only car with `sale_price_usd` \$1 and \$350000 will be kept, remaining will be removed from the dataframe `auto.csv`.

In [21]:
autos = autos[autos['sale_price_usd'].between(1, 351000, 1000)]

In [22]:
autos.sale_price_usd.describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: sale_price_usd, dtype: float64

In [23]:
autos['odometer_km'].value_counts().sort_index(ascending=False).head(20)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64

In [24]:
autos['odometer_km'].value_counts().sort_index(ascending=True).head(20)

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Values in `odometer_km` column seem to be realistic and good to go.

In [25]:
# calculating the percentage distribution of values in `date_crawled` column

(autos['date_crawled']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The site was crawled daily over almost a one-month period during March and April 2016. The distribution of listings crawled on each day is uniform.

In [26]:
# calculating the percentage distribution of values in `ad_created` column

(autos['ad_created']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index(ascending=False)
)

2016-04-07    0.001256
2016-04-06    0.003253
2016-04-05    0.011819
2016-04-04    0.036858
2016-04-03    0.038855
2016-04-02    0.035149
2016-04-01    0.033687
2016-03-31    0.031875
2016-03-30    0.033501
2016-03-29    0.034037
2016-03-28    0.034984
2016-03-27    0.030989
2016-03-26    0.032266
2016-03-25    0.031751
2016-03-24    0.029280
2016-03-23    0.032060
2016-03-22    0.032801
2016-03-21    0.037579
2016-03-20    0.037949
2016-03-19    0.033687
2016-03-18    0.013590
2016-03-17    0.031278
2016-03-16    0.030125
2016-03-15    0.034016
2016-03-14    0.035190
2016-03-13    0.017008
2016-03-12    0.036755
2016-03-11    0.032904
2016-03-10    0.031895
2016-03-09    0.033151
                ...   
2016-02-21    0.000062
2016-02-20    0.000041
2016-02-19    0.000062
2016-02-18    0.000041
2016-02-17    0.000021
2016-02-16    0.000021
2016-02-14    0.000041
2016-02-12    0.000041
2016-02-11    0.000021
2016-02-09    0.000021
2016-02-08    0.000021
2016-02-07    0.000021
2016-02-05 

The ads were created at a huge scale specially after Feb 2016, but few ads are more than 6 months old.

In [27]:
# calculating the percentage distribution of values in `last_seen` column

(autos['last_seen']
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index(ascending=False)
)

2016-04-07    0.131947
2016-04-06    0.221806
2016-04-05    0.124761
2016-04-04    0.024483
2016-04-03    0.025203
2016-04-02    0.024915
2016-04-01    0.022794
2016-03-31    0.023783
2016-03-30    0.024771
2016-03-29    0.022341
2016-03-28    0.020859
2016-03-27    0.015649
2016-03-26    0.016802
2016-03-25    0.019211
2016-03-24    0.019767
2016-03-23    0.018532
2016-03-22    0.021373
2016-03-21    0.020632
2016-03-20    0.020653
2016-03-19    0.015834
2016-03-18    0.007351
2016-03-17    0.028086
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-14    0.012602
2016-03-13    0.008895
2016-03-12    0.023783
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-08    0.007413
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

The `last_seen` crawler recorded the date it last saw any listing and it allows us to determine the removed listing more likely because the cars were sold. Moreover, in the last 3 months the `last_seen` values picked a hype in comparison to past months which might be due to high sales spike (highly unlikely) or due to the crawling ending period.

In [28]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We needed to look further into registration_year column and found something odd as it shows maximum value as 9999 which is inaccurate and also minimum value as 1000, even before the invention of the car.

In [29]:
# determining the highest and lowest acceptable values for 
# the `registration_year` column

autos.registration_year.value_counts(normalize=True)

2000    0.064985
2005    0.060455
1999    0.059652
2004    0.055657
2003    0.055575
2006    0.054978
2001    0.054278
2002    0.051189
1998    0.048656
2007    0.046886
2008    0.045609
2009    0.042932
1997    0.040173
2011    0.033419
2010    0.032719
2017    0.028663
1996    0.028271
2012    0.026974
1995    0.025265
2016    0.025121
2013    0.016535
2014    0.013652
1994    0.012952
2018    0.009678
1993    0.008751
2015    0.008072
1992    0.007619
1990    0.007145
1991    0.006980
1989    0.003583
          ...   
1950    0.000062
9999    0.000062
1800    0.000041
1951    0.000041
1941    0.000041
1934    0.000041
1954    0.000041
1957    0.000041
1955    0.000041
2019    0.000041
1001    0.000021
1939    0.000021
4500    0.000021
1953    0.000021
1111    0.000021
4800    0.000021
5911    0.000021
1943    0.000021
1938    0.000021
1929    0.000021
2800    0.000021
6200    0.000021
4100    0.000021
8888    0.000021
1927    0.000021
9000    0.000021
1948    0.000021
1000    0.0000

*Lowest acceptable* value will be 1900 and *highest acceptable* value will be 2016, because the first car was invented in 1885 and it took time for the car registration to have a legal effect and that happended in first few decade of 1900s.

In [30]:
# determining the percentage of outliers in `registration_year` column

(~autos['registration_year']
 .between(1900, 2016)).sum() / (autos.registration_year
                                .between(1900, 2016)).sum()

0.040359032582849556

Above calculated percentage is 4% of series `registration_year`, we will remove these rows from the dataframe.

In [31]:
# removing the values outside the upper and lower bounds and calculating
# distribution of frequencies

autos = (autos[(autos.registration_year > 1900) |
                         (autos.registration_year <= 2016)]
)

autos.registration_year.value_counts(normalize=True).head(10)

2000    0.064985
2005    0.060455
1999    0.059652
2004    0.055657
2003    0.055575
2006    0.054978
2001    0.054278
2002    0.051189
1998    0.048656
2007    0.046886
Name: registration_year, dtype: float64

Above output depicts that most registrations of cars happened in the past two decades.

In [32]:
# exploring unique values in the `brand` column to decide on the brand 
# to use for aggregation

autos.car_brand.value_counts(normalize=True)

volkswagen        0.212828
opel              0.108658
bmw               0.108597
mercedes_benz     0.095789
audi              0.085823
ford              0.069639
renault           0.047874
peugeot           0.029445
fiat              0.025986
seat              0.018944
skoda             0.016061
nissan            0.015258
mazda             0.015217
smart             0.014290
citroen           0.014125
toyota            0.012581
hyundai           0.009945
sonstige_autos    0.009698
volvo             0.009039
mini              0.008607
mitsubishi        0.008216
honda             0.007989
kia               0.007104
alfa_romeo        0.006610
porsche           0.005910
suzuki            0.005889
chevrolet         0.005663
chrysler          0.003480
dacia             0.002656
daihatsu          0.002512
jeep              0.002224
subaru            0.002121
land_rover        0.002039
saab              0.001627
daewoo            0.001565
jaguar            0.001524
trabant           0.001400
r

Volkswagen is the most famous brand among top 5 brands, however, remaining 4 are of German manufacturers.

In [33]:
# extracting brand names having distribution freq of more than 5%
famous_brands = (autos.car_brand.value_counts(normalize=True)
 [autos.car_brand.value_counts(normalize=True) > .05].index
)

famous_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [34]:
# aggregating data

mean_price = {}
for b in famous_brands:
    mean = (autos.loc
                  [autos.car_brand == b, 'sale_price_usd']
                  .mean()
                   )
    mean_price[b] = int(mean)

print(mean_price)    

{'volkswagen': 5332, 'opel': 2944, 'bmw': 8261, 'mercedes_benz': 8536, 'audi': 9212, 'ford': 3728}


Looking at above results, it is obvious that *Ford* and *Opel* are less expensive cars; whereas *Audi*, *Mercedes_benz* and *BMW* are more expensive ones while *Volkswagen* is in between these two groups.

In [35]:
# calculating mean mileag for each of the top car brands

mileage_km = {}
for b in famous_brands:
    mean_mileage = (autos.loc
                    [autos.car_brand == b, 'odometer_km']
                    .mean()
                   )
    mileage_km[b] = int(mean_mileage)

print(mileage_km)

{'volkswagen': 128896, 'opel': 129383, 'bmw': 132682, 'mercedes_benz': 130796, 'audi': 129492, 'ford': 124349}


In [36]:
price_series = pd.Series(mean_price).sort_values(ascending=False)
price_series

audi             9212
mercedes_benz    8536
bmw              8261
volkswagen       5332
ford             3728
opel             2944
dtype: int64

In [37]:
mileage_series = pd.Series(mileage_km).sort_values(ascending=False)
mileage_series

bmw              132682
mercedes_benz    130796
audi             129492
opel             129383
volkswagen       128896
ford             124349
dtype: int64

In [38]:
price_mileage_dataframe = pd.DataFrame(price_series, 
                                       columns=['mean_price'])

price_mileage_dataframe

mean_price
audi                 9212
mercedes_benz        8536
bmw                  8261
volkswagen           5332
ford                 3728
opel                 2944

In [39]:
price_mileage_dataframe['mean_mileage'] = mileage_series

price_mileage_dataframe

mean_price  mean_mileage
audi                 9212        129492
mercedes_benz        8536        130796
bmw                  8261        132682
volkswagen           5332        128896
ford                 3728        124349
opel                 2944        129383

Based on the above resulted dataframe, it may be concluded that there is a slight trend to the less expensive vehicles having lower mileage, with the more expensive vehicles having higher mileage.